File: Cpp/Easy/ReverseLinkedList.cpp
/*****************************************************
 *
 *  206. Reverse Linked List solution c++
 *
 *  Using Stack 
 *
 *  Joel M. Gonzalez Rodriguez
 *
*/****************************************************

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

ListNode *reverseList(ListNode *head)
{
    ListNode *current = head;
    stack<int> myStack;

    while (current != nullptr)
    {
        myStack.push(current->val);
        current = current->next;
    }

    int size = myStack.size();
    current = head;

    for (int i = 0; i < size; i++)
    {
        current->val = myStack.top();
        myStack.pop();
        current = current->next;
    }

    return head;
}

File: Python/Easy/Reverse_Linked_List.py
'''
    
    206. Reverse Linked List

    Solution:

'''

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return head

        prev_node = head
        head = head.next
        prev_node.next = None
        
        while head != None:
            new_node = ListNode()
            new_node.val = head.val
            new_node.next = prev_node
            prev_node = new_node

            head = head.next
            
        return prev_node